### PR TITLE
DAOS-11305 engine: add mutex in sched_xstream_stopping() and dss_xstreams_fini()

### PR DESCRIPTION
**Pull request by wiliamhuang**
To avoid possible race condition between dss_xstreams_fini and sched_create_thread.

### CODE REVIEW

File: src/engine/srv.c [hunk 1]
```diff
@@ -800,17 +800,20 @@ dss_xstreams_fini(bool force)
...
 
+	ABT_mutex_lock(xstream_data.xd_mutex);
 	/* Notify all xstreams to reject new ULT creation first */
 	for (i = 0; i < xstream_data.xd_xs_nr; i++) {
 		dx = xstream_data.xd_xs_ptrs[i];
 		if (dx == NULL)
 			continue;
 		ABT_future_set(dx->dx_stopping, dx);
 	}
+	ABT_cond_broadcast(xstream_data.xd_ult_barrier);
```
Comment (liuxuezhao) [inline src/engine/srv.c:812]: why signal xd_ult_barrier here?

File: src/engine/srv.c [hunk 2]
```diff
@@ -800,17 +800,20 @@ dss_xstreams_fini(bool force)
...
+	ABT_mutex_unlock(xstream_data.xd_mutex);
+
+	rc = bio_nvme_ctl(BIO_CTL_NOTIFY_STARTED, &started);
```
Comment (liuxuezhao) [inline src/engine/srv.c:815]: and not sure what's this call's purpose?

BTW, may need to confirm with the person who reported the ticket if the problem is resolved. Thx.

**Review by NiuYawei — CHANGES_REQUESTED** [review]
There was no race, and this mutex will introduce contention and kill the performance.

File: src/engine/srv.c (hunk 1 above)
Comment (wiliamhuang) [reply src/engine/srv.c:812]: Thank you! This was added by a mistake. I am going to fix it.

File: src/engine/srv.c (hunk 2 above)
Comment (wiliamhuang) [reply src/engine/srv.c:815]: @NiuYawei Could you please shed some light about the Xuezhao's question?
I will check with the user to see whether the problem is resolved or not. Thank you!

Comment (NiuYawei) [reply src/engine/srv.c:815]: I don't think there is any race in current implementation, please check my comment on Jira ticket.

Comment (wiliamhuang) [reply src/engine/srv.c:815]: @NiuYawei ok. I will check it out. Thank you very much!

**Pull request closed without merging**